Параллельная индексация документов

MedAlexey opened the pull request.

MedAlexey committed:
Утилита доработана таким оразом, чтобы индексация документов происходила параллельно.

tariel-x commented:
Введите, пожалуйста, какой-нибудь тип вроде `type Index map[string]map[string]int` внутри пакета с индексом, чтоб код был наглядней.

tariel-x commented:
В прошлый раз не заметил: разделите, пожалуйста, коммуникацию с "внешним миром" (ФС, пользовательский интерфейс) и работу с индексом на разные пакеты. То есть пакет-индексатор должен содержать только код, работающий со структурой и данными индекса: добавление новых текстов, поиск по индексу. А другой пакет (main или отдельный) - всю остальную работу. Так проще будет впоследствии заменить взаимодействие через консоль на взаимодействие по HTTP, а чтение из базы - на чтение страниц или из БД.

tariel-x commented:
В коде есть race:
```
$ go run main.go books/hard books/noon books/prisoners books/time 
fatal error: concurrent map read and map write

goroutine 20 [running]:
runtime.throw(0x4cdcb9, 0x21)
        /usr/local/go/src/runtime/panic.go:608 +0x72 fp=0xc000074e10 sp=0xc000074de0 pc=0x427c72
runtime.mapaccess2_faststr(0x4ae2a0, 0xc000068150, 0xc0006bc711, 0x8, 0xc000cb1680, 0x1)
        /usr/local/go/src/runtime/map_faststr.go:110 +0x458 fp=0xc000074e80 sp=0xc000074e10 pc=0x4101a8
main.main.func1(0xc000068150, 0x7ffd2cddeb11, 0xf, 0xc000084010, 0xc000084020)
        /home/nikita/go/src/github.com/polis-mail-ru-golang-1/t2-invert-index-search-MedAlexey/main.go:33 +0x205 fp=0xc000074fb8 sp=0xc000074e80 pc=0x496895
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc000074fc0 sp=0xc000074fb8 pc=0x451631
created by main.main
        /home/nikita/go/src/github.com/polis-mail-ru-golang-1/t2-invert-index-search-MedAlexey/main.go:28 +0x182

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc000084018)
        /usr/local/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc000084010)
        /usr/local/go/src/sync/waitgroup.go:130 +0x64
main.main()
        /home/nikita/go/src/github.com/polis-mail-ru-golang-1/t2-invert-index-search-MedAlexey/main.go:48 +0x1ac

goroutine 19 [runnable]:
main.main.func1(0xc000068150, 0x7ffd2cddeb06, 0xa, 0xc000084010, 0xc000084020)
        /home/nikita/go/src/github.com/polis-mail-ru-golang-1/t2-invert-index-search-MedAlexey/main.go:35 +0x2ca
created by main.main
        /home/nikita/go/src/github.com/polis-mail-ru-golang-1/t2-invert-index-search-MedAlexey/main.go:28 +0x182
exit status 2
```

Попробуйте индексировать большие файлы, например, отсюда: https://github.com/polis-mail-ru-golang-1/examples/tree/master/t2/books

MedAlexey committed:
> внутри пакета с индексом введён новый тип (type Index map[string]map[string]int) для наглядности кода
> коммуникация с "внешним миром" выведена из пакета с индексом в пакет main
> rece устранена

tariel-x commented:
У вас определён тип `makeInvertIndex.Index`, вы его можете использовать по всему проекту вместо `map[string]map[string]int`

tariel-x commented:
:+1:
